fix: ignore ENOENT in `injectSourcesContent`

aleclarson opened the pull request:
### Description

This is important for virtual modules and malformed source maps, both of which can throw `ENOENT` errors when injecting `sourcesContent` into source maps.

### Solution

Expect sourcemaps for virtual modules to either contain a pre-populated `sourcesContent` array or have a `sources` array of **absolute paths only**. Otherwise, a warning will be printed (at most one time).

Expect sourcemaps for normal modules to either contain a pre-populated `sourcesContent` array or have a `sources` array that never points to non-existent files. Relative source paths are allowed, and they're resolved with `sourceRoot` (if defined) or the parent directory of the module. When a non-existent source is found, a warning will printed (at most once per sourcemap for the lifetime of the Vite server).

---

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [x] Bug fix
- [ ] New Feature
- [ ] Documentation update
- [ ] Other

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/.github/contributing.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/.github/contributing.md#pull-request-guidelines) and follow the [Commit Convention](https://github.com/vitejs/vite/blob/main/.github/commit-convention.md).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [ ] Ideally, include relevant tests that fail without this PR but pass with it.

aleclarson committed:
fix: ignore ENOENT in `injectSourcesContent`

This is important for virtual modules and malformed source maps, both of which can throw ENOENT errors when injecting `sourcesContent`.

patak-js commented:
Would it be possible to differentiate between malformed source maps and virtual modules so we can emit a warning when the maps are malformed instead of a silent ignore? Also, do you think it is possible to add a test case?

Shinigami92 commented:
Is this `p2-nice-to-have` or `p3-minor-bug`?

aleclarson commented:
> Would it be possible to differentiate between malformed source maps and virtual modules so we can emit a warning when the maps are malformed instead of a silent ignore?

Maybe, should we bring this up with Evan?

raythurnevoid commented on packages/vite/src/node/server/sourcemap.ts:
```diff
@@ -5,16 +5,26 @@ export async function injectSourcesContent(
+          'utf-8'
+        )
+      } catch (e) {
+        if (e.code == 'ENOENT') return
```
Maybe here it's better to log some warning, packages authors should fix this problem by not including sourcemaps in packages.

aleclarson replied:
No, sourcemaps should not be discouraged. Sources should be published by npm packages, for easier debugging. The `ENOENT` errors should only occur for missing files and "virtual files" (intermediate files created by plugins). It makes sense to ignore errors for virtual files, but the question is how to differentiate missing files from virtual ones.

aleclarson commented on packages/vite/src/node/server/sourcemap.ts:
```diff
@@ -5,16 +5,26 @@ export async function injectSourcesContent(
-      )
+      try {
+        map.sourcesContent![i] = await fs.readFile(
+          path.resolve(sourceRoot, decodeURI(sourcePath)),
```
I think we don't need to catch `ENOENT` errors here, since it's not possible for a virtual module to be in the `sources` of a real file's sourcemap. Therefore, we can know that failed `fs.readFile` calls are related to a malformed source path.

aleclarson commented on packages/vite/src/node/server/sourcemap.ts:
```diff
@@ -5,16 +5,26 @@ export async function injectSourcesContent(
+      path.resolve(path.dirname(file), map.sourceRoot || '')
+    )
+  } catch (e) {
+    if (e.code == 'ENOENT') return
```
I don't think we need to worry about differentiating between virtual modules and missing files here. We can just assume the `file` argument represents a virtual module if `fs.realpath` fails. So swallowing the `ENOENT` error is enough here.

aleclarson replied:
One implication of just ignoring the error here: Virtual modules need to add `sourcesContent` to their sourcemaps manually. If we want to avoid that requirement, we can assume `sources` in virtual modules' sourcemaps contains absolute file paths, or at least resolve them relative to the project root.

aleclarson committed:
fix: avoid swallowing ENOENT errors for malformed source paths

Also, when the `file` is a virtual module, assume source paths are already absolute.

aleclarson committed:
fix: throw a clearer error for non-existent sources

dimfeld commented:
Am I correct that with the changes to this PR it now throws errors once again in the "malformed sourcemap" case, as it currently does? I think it would be very helpful if malformed source maps in dependencies, whether from bad paths or just failure to package the source with the NPM package, didn't completely kill the build. 

This is an especially sensitive case IMO because it mostly comes up with dependencies where there's a much longer turnaround to get the source maps fixed.

aleclarson commented:
@dimfeld That's a good point. It's not a fatal error, so we shouldn't kill the build. 👍

aleclarson committed:
feat: warn once per sourcemap with missing source files

aleclarson committed:
fix(sourcemap): ensure `sources` and `sourcesContent` are the same length

aleclarson commented:
@patak-js @antfu This PR has been reworked. Please read the "Solution" section in the OP, and review the code. Thx

HerringtonDarkholme commented:
Hi @patak-js, I wonder if this PR could be released to a new version of Vite soon? Thanks!

patak-js commented:
We are doing one release per week, it will be out next tuesday

wmzy commented:
Hi @aleclarson and @patak-js ,
I also encountered this bug.
I debug and found the `transformResult.map.sourcesContent` will be excluded always:
https://github.com/vitejs/vite/blob/828a8e538ce289922b9301339be4d5d6ea84b496/packages/vite/src/node/server/transformRequest.ts#L131-L152
The `transform` will call https://github.com/vitejs/vite/blob/828a8e538ce289922b9301339be4d5d6ea84b496/packages/vite/src/node/utils.ts#L434 to combine sourcemap which will call `remapping` with `excludeContent` option as `true`.

patak-js commented:
@wmzy do you see an issue then with the way this PR is implemented? Could you create a bug report if that is the case?

wmzy commented:
@patak-js this PR fixed the bug in `injectSourcesContent `. 
But in my case, the `injectSourcesContent` shoud not been call if the `pluginContainer.transform` has not lost the `sourcesContent`.

I will try a PR to optimize this. It's hard to explain with my poor English.

wmzy commented:
@patak-js [The PR](https://github.com/vitejs/vite/pull/4362) has been created, could you review it?
